common/metrics/client.go

```diff
@@ -104,7 +107,14 @@ func (m *ClientImpl) UpdateGauge(scopeIdx int, gaugeIdx int, value float64) {
 // Scope return a new internal metrics scope that can be used to add additional
 // information to the metrics emitted
 func (m *ClientImpl) Scope(scopeIdx int, tags ...Tag) Scope {
-	return newMetricsScope(m.childScopes[scopeIdx], m.metricDefs, false).Tagged(tags...)
+	strScope := stringifyScope(scopeIdx, tags...)
+	scope, ok := m.cachedScopes[strScope]
+	if ok {
+		return scope
+	}
+	scope = newMetricsScope(m.childScopes[scopeIdx], m.metricDefs, false).Tagged(tags...)
+	m.cachedScopes[strScope] = scope
```

Comment:
You might want to use a bounded size map here just to be defensive - eventhough the size itself can be rather large. When the size is exceeded, some simple strategy such as evicting at random is prolly fine to start with. Iff the size is chosen carefully, no eviction is likely expected.